packages/Thoth.Json.Core/Encode.fs

```diff
@@ -190,44 +290,26 @@ module Encode =
 
     module Enum =
 
-        let byte<'TEnum when 'TEnum: enum<byte>> (value: 'TEnum) : Json =
+        let byte<'TEnum when 'TEnum: enum<byte>> (value: 'TEnum) : IEncodable =
             LanguagePrimitives.EnumToValue value |> byte
 
-        let sbyte<'TEnum when 'TEnum: enum<sbyte>> (value: 'TEnum) : Json =
+        let sbyte<'TEnum when 'TEnum: enum<sbyte>> (value: 'TEnum) : IEncodable =
             LanguagePrimitives.EnumToValue value |> sbyte
 
-        let int16<'TEnum when 'TEnum: enum<int16>> (value: 'TEnum) : Json =
+        let int16<'TEnum when 'TEnum: enum<int16>> (value: 'TEnum) : IEncodable =
             LanguagePrimitives.EnumToValue value |> int16
 
-        let uint16<'TEnum when 'TEnum: enum<uint16>> (value: 'TEnum) : Json =
+        let uint16<'TEnum when 'TEnum: enum<uint16>> (value: 'TEnum) : IEncodable =
             LanguagePrimitives.EnumToValue value |> uint16
 
-        let int<'TEnum when 'TEnum: enum<int>> (value: 'TEnum) : Json =
+        let int<'TEnum when 'TEnum: enum<int>> (value: 'TEnum) : IEncodable =
             LanguagePrimitives.EnumToValue value |> int
 
-        let uint32<'TEnum when 'TEnum: enum<uint32>> (value: 'TEnum) : Json =
+        let uint32<'TEnum when 'TEnum: enum<uint32>> (value: 'TEnum) : IEncodable =
             LanguagePrimitives.EnumToValue value |> uint32
 
-    let option (encoder: 'a -> Json) =
+    let option (encoder: Encoder<'a>) =
         Option.map encoder >> Option.defaultWith (fun _ -> nil)
 
-    let rec toJsonValue (helpers: IEncoderHelpers<'JsonValue>) (json: Json) =
-        match json with
-        | Json.String value -> helpers.encodeString value
-        | Json.IntegralNumber value -> helpers.encodeIntegralNumber value
-        | Json.Object values ->
-            let o = helpers.createEmptyObject ()
-
-            values
-            |> Seq.iter (fun (k, v) ->
-                helpers.setPropertyOnObject (o, k, toJsonValue helpers v)
-            )
-
-            o
-        | Json.Char value -> helpers.encodeChar value
-        | Json.DecimalNumber value -> helpers.encodeDecimalNumber value
-        | Json.Null -> helpers.encodeNull ()
-        | Json.Boolean value -> helpers.encodeBool value
-        | Json.Array value ->
-            value |> Array.map (toJsonValue helpers) |> helpers.encodeArray
-        | Json.Unit -> helpers.encodeNull ()
+    let rec toJsonValue (helpers: IEncoderHelpers<'JsonValue>) (json: IEncodable) =
+        json.Encode(helpers)
```

Comment:
I am wondering if this function still make sense now.

If we keep it we should remove the `rec` keyword and we can probably `inline` it.